server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -169,14 +180,19 @@ protected WritePrimaryResult<ShardUpsertRequest, ShardResponse> processRequestIt
             //  Would need to have [x, y, ...]
             assert request.insertColumns() == null || insertColumns.subList(0, request.insertColumns().length).equals(Arrays.asList(request.insertColumns()))
                 : "updateToInsert.columns() must be a superset of insertColumns where the start is an exact overlap. It may only add new columns at the end";
-        } else {
-            insertColumns = Arrays.asList(request.insertColumns());
         }
+        // While resolving references we must always use a fresh table instance as the current one can get stale
+        // when columns are added dynamically.
+        Function<ColumnIdent, Reference> getRef = c -> {
+            DocTableInfo table = schemas.getTableInfo(relationName, Operation.READ);
+            return table.getReference(c);
+        };
         Indexer indexer = new Indexer(
```

Comment:
Do I get it right that table is resolved per inserted item/per row if we insert into table with object column?

As stated in the first commit's message, updating targets doesn't help because 
> `AddColumnRequest` would not return these references as they already exists.

I wonder if making `AddColumnResponse` return all request targets (not only non-existing) would help to avoid table refreshing since it's an atomic update and potentially can return fresh state.

Reply:
Yes, the table is resolved on each item. But the table is cached so it should not have any performance impacts.

> I wonder if making AddColumnResponse return all request targets (not only non-existing) would help to avoid resolving table per item since it's an atomic update and is capable of returning fully fresh state?

I think, resolving a most-recent table is the more lightweight and safe operation as it reflects the actual state the current node is operating on.

Reply:
> I also have the impression that it's now a bit duplicated. We pass along the added columns but also the option to re-resolve the reference. Is both necessary? And why?

The are race-conditions with concurrent `AddColumnRequest`, one request may be processed while another already triggered an new column creation. In this case the response will not contain these columns but the most-recent table does as it already processed the new cluster state.

Reply:
Also valid point, I look into this again.

Reply:
As we need to replace the `ValueIndexer` when a column was created (we need the OID), the new value indexer, especially the `ObjectIndexer` must be able to resolve the childs references.

I dived into this again and I think this can be done by one of the following solutions:

- a) the `getRef` will always use a most-recent table, in this case we do not need the `addedColumns`.

- b) we wrap the `getRef` to look into the `addedColumns` first as the current table instance is stale. We could either pass in a new `getRef` (wrapper) or let each value indexer (actually only the object indexer) create one itself.
  In this case, the `AddColumnTask` must always return the requested references, even when they already exists.

The current solution does somehow both as requested to-add columns are not always returned.

I slightly favour option b) as this is a bit easier to handle in tests and we can avoid requesting a most-recent table instance on every lookup.

@mfussenegger wdyt?

Reply:
Is b) safe in a mixed cluster scenario? I think we can't be sure that the add-column action would return all columns.

Reply:
I've pushed a fixup commit which removes the `addedColumns` from the `updateTargets` API and pass in a new `getRef` using the most-recent `DocTableInfo` instead (https://github.com/crate/crate/pull/14726/commits/b0636c4ca20efbad80670819f6af5da5d9f61065)
Additionally I've removed the first commit of this PR (https://github.com/crate/crate/commit/a3367a92b33f4b5d3ac1719d8d231c5c32eed0a2) as we do not need a most-recent table info on each lookup anymore.